caikit/runtime/service_generation/data_stream_source.py

```diff
@@ -52,124 +57,74 @@
+    source type.
+    """
 
-    def __getstate__(self) -> bytes:
-        """A DataStreamSource is pickled by serializing its source
-        representation. This is particularly useful when sharing data streams
-        across subprocesses to run training in an isolated process.
+    def __init__(self, config: aconfig.Config, instance_name: str):
+        """Construct with the basic factor constructible interface and store the
+        args for use by the child
         """
-        return self.to_binary_buffer()
+        self._config = config
+        self._instance_name = instance_name
 
-    def __setstate__(self, pickle_bytes: bytes):
-        """Unpickling a DataStreamSource basically involves unpacking the
-        serialized source representation. The catch is that the oneof is
-        represented strangely in __dict__, so we need to explicitly set all
-        oneof members.
+    ## Abstract Interface ##
+
+    @abc.abstractmethod
+    def get_stream_message_type(self, element_type: type) -> Type[DataBase]:
+        """Get the type of the dataobject class that will be used as the source
+        information
         """
-        new_inst = self.__class__.from_binary_buffer(pickle_bytes)
-        setattr(self, new_inst.which_oneof("data_stream"), new_inst.data_stream)
-        self.generator_func = self._generator
-        self.generator_args = tuple()
-        self.generator_kwargs = {}
 
-    # pylint: disable=too-many-return-statements
-    def to_data_stream(self) -> DataStream:
-        """Convert to the target data stream type based on the source type"""
+    @abc.abstractmethod
+    def to_data_stream(
+        self, source_message: Type[DataBase], element_type: type
+    ) -> DataStream:
+        """Convert an instance of the source message type into a DataStream"""
 
-        # Determine which of the value types is set
-        set_field = None
-        for field_name in self.get_proto_class().DESCRIPTOR.fields_by_name:
-            if getattr(self, field_name) is not None:
-                assert (
-                    set_field is None
-                ), "Found DataStreamSource with multiple sources set"
-                set_field = field_name
+    @abc.abstractmethod
+    def get_field_number(self) -> int:
+        """Allow plugins to return a static field number"""
 
-        # If no field is set, return an empty DataStream
-        if set_field is None:
-            log.debug3("Returning empty data stream")
-            return DataStream.from_iterable([])
+    ## Public Methods ##
 
-        # If a S3 pointer is given, raise not implemented
-        if set_field == "s3files":
-            error(
-                "<COR80419785E>",
-                NotImplementedError(
-                    "S3Files are not implemented as stream sources in this runtime."
-                ),
-            )
+    def get_field_name(self, element_type: type) -> str:
```

Comment:
I was really torn on this one. One the one hand, I would prefer if this were a class property, or at the very least an instance property and NOT dependent on the `element_type`, but on the other hand, the default implementation needs to know the name of the stream message which _is_ dependent on the `element_type`.

Reply:
🤔  but the result of this should never include any details about the element type?

i.e. it should just be `file`, `jsondata`, `directory`...

Maybe worth some more huddle time and thinking

Reply:
I _could_ imagine a circumstance where you would want to have the element type inform this, right? Not sure it's a good idea, but I could imagine it.